homeassistant/components/kulersky/__init__.py

```diff
@@ -1,33 +1,126 @@
+from homeassistant.helpers.discovery_flow import DiscoveryKey
 
-from .const import DATA_ADDRESSES, DATA_DISCOVERY_SUBSCRIPTION, DOMAIN
+from .const import DOMAIN
 
 PLATFORMS = [Platform.LIGHT]
 
+_LOGGER = logging.getLogger(__name__)
+
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Kuler Sky from a config entry."""
-    if DOMAIN not in hass.data:
-        hass.data[DOMAIN] = {}
-    if DATA_ADDRESSES not in hass.data[DOMAIN]:
-        hass.data[DOMAIN][DATA_ADDRESSES] = set()
-
+    ble_device = async_ble_device_from_address(
+        hass, entry.data[CONF_ADDRESS], connectable=True
+    )
+    if not ble_device:
+        raise ConfigEntryNotReady(
+            translation_domain=DOMAIN,
+            translation_key="cannot_connect",
+        )
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
     return True
 
 
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload a config entry."""
-    # Stop discovery
-    unregister_discovery = hass.data[DOMAIN].pop(DATA_DISCOVERY_SUBSCRIPTION, None)
-    if unregister_discovery:
-        unregister_discovery()
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
 
-    hass.data.pop(DOMAIN, None)
 
-    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+async def async_migrate_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Migrate old entry."""
+    _LOGGER.debug("Migrating from version %s", config_entry.version)
+
+    # Version 1 was a single entry instance that started a bluetooth discovery
+    # thread to add devices. Version 2 has one config entry per device, and
+    # supports core bluetooth discovery
+    if config_entry.version == 1:
+        dev_reg = dr.async_get(hass)
+        entity_reg = er.async_get(hass)
+        devices = dev_reg.devices.get_devices_for_config_entry_id(config_entry.entry_id)
+
+        if len(devices) == 0:
+            _LOGGER.error("Unable to migrate; No devices registered")
+            return False
+
+        first_device = devices[0]
+        domain_identifiers = [i for i in first_device.identifiers if i[0] == DOMAIN]
+        address = next(iter(domain_identifiers))[1]
+        hass.config_entries.async_update_entry(
+            config_entry,
+            title=first_device.name or address,
+            data={CONF_ADDRESS: address},
+            discovery_keys=MappingProxyType(
+                {
+                    BLUETOOTH_DOMAIN: (
+                        DiscoveryKey(
+                            domain=BLUETOOTH_DOMAIN,
+                            key=address,
+                            version=1,
+                        ),
+                    )
+                }
+            ),
+            unique_id=address,
+            version=2,
+        )
+
+        # Create new config entries for the remaining devices
```

Comment:
We usually don't create `ConfigEntry` directly and let config flows do this.

Maybe make these import flows to create the new entries instead

Reply:
I think if we want to keep user involved then we could just simply this migration by dropping everything below the comment. Any remaining devices will get picked up by default bluetooth discovery.

Reply:
I think its fine to add them but we should avoid creating `ConfigEntry` objects in integrations and let the config flow tooling take care of all that.

```
			hass.async_create_task(
                hass.config_entries.flow.async_init(
                    DOMAIN,
                    context={"source": config_entries.SOURCE_INTEGRATION_DISCOVERY},
                    data=`....data needed to import it....`,
                ),
            )
```

Than implement `async_step_integration_discovery` in the config flow.